Trips

flacks opened the pull request:
This is a working implementation of the Trips concept, with:
- A new Trip model
- An update to the User model to link a list of trips associated with a user (as a rider) via a "Riders" Join Table
- A Trip controller with a `/trips` endpoint which returns `ResponseEntity`'s for `GET`, `POST`, `PUT`, and `DELETE` HTTP requests
- A Trip repository with 2 custom methods for fetching trips by rider/driver IDs
- A Trip service interface with it's respective implementation

grturner committed:
Model Updates: Users & Addresses Created Address model, and refactored Users and DistanceServiceImpl to update

Signed-off-by: Gilbert Turner <grturner@5x5code.com>

flacks committed:
pom.xml updates

- Update Spring Boot version
- Update Oracle JDBC dependency
- Add Lombok
- Add version to Google's json-simple

flacks committed:
WIP: Trip implementation

Co-authored-by: Jose Avalos <jl.avalos10@gmail.com>

jlavalos73 committed:
Add Trip service implementation

flacks committed:
Finish working Trips concept

- Return ResponseEntity's at the Trip controller level
- Bugfix Trip model
- Add missing @Repository annotation to Trip repository, use native SQL
  for Join Table lookup on getTripsByRiderId

Co-authored-by: The Camel <jcfo233@gmail.com>
Co-authored-by: Jose Avalos <jlavalos73@gmail.com>
Co-authored-by: Galo Romero <grome014@fiu.edu>
Co-authored-by: Gilbert Turner <grturner@5x5code.com>

flacks committed:
Merge branch 'dev' into trips

flacks committed:
Remove unused code, tidy up imports

IkenoXamos commented on pom.xml:
```diff
@@ -1,9 +1,9 @@
-  <artifactId>rideshare-user-service</artifactId>
-  <version>0.0.1-SNAPSHOT</version>
-  
+	<modelVersion>4.0.0</modelVersion>
```
Was any of this actually necessary?

IkenoXamos commented on pom.xml:
```diff
@@ -60,100 +59,91 @@
-		<!-- login credentials moved to Jenkins build -->
+		<sonar.organization>batch-738-mathew-oberlies</sonar.organization>
+		<sonar.host.url>https://sonarcloud.io</sonar.host.url>
+    <sonar.coverage.exclusion>com/revature/models/*.java</sonar.coverage.exclusion>
```
You spent time fixing alignment in this file, and this one line was changed to be misaligned.

IkenoXamos commented on pom.xml:
```diff
@@ -60,100 +59,91 @@
+			<scope>test</scope>
+		</dependency>
+
+<!--		<dependency>-->
```
Single line comments necessary over multi-line comments?

flacks committed:
Fix indentation

IkenoXamos commented on src/main/java/com/revature/controllers/TripController.java:
```diff
@@ -0,0 +1,140 @@
+     */
+    @ApiOperation(value = "Updates a trip by ID", tags = {"Trip"})
+    @PutMapping
+    public ResponseEntity<Trip> updateTrip(@Valid @RequestBody Trip trip) {
```
I used this structure in previous demos, and it works alright, but it has a critical flaw: If you receive a bad request, you save it as a new entity. This is most likely not desired behavior.

flacks replied:
Thank you for pointing this out. Resolved.

IkenoXamos commented on src/main/java/com/revature/controllers/TripController.java:
```diff
@@ -0,0 +1,140 @@
+     */
+    @ApiOperation(value = "Adds a new trip")
+    @PostMapping
+    public ResponseEntity<Trip> addTrip(@Valid @RequestBody Trip trip) {
```
This has the same flaw: You create entities when it is a bad request.

flacks replied:
Resolved.

IkenoXamos commented on src/main/java/com/revature/controllers/TripController.java:
```diff
@@ -0,0 +1,140 @@
+    }
+
+    /**
+     * HTTP DELETE method (/cars/{id})
```
This might have been copy-pasted, but the javadoc comment says /cars/id instead of /trips/id.

flacks replied:
Fixed.

IkenoXamos commented on src/main/java/com/revature/models/Trip.java:
```diff
@@ -0,0 +1,43 @@
+@Component
+@Entity
+@Table(name = "trips")
+@Getter @NoArgsConstructor @AllArgsConstructor @Setter @EqualsAndHashCode @ToString
```
Preferably use @ Data annotation for consistency with other models.

flacks replied:
Thank you! Fixed.

IkenoXamos commented on src/main/java/com/revature/repositories/TripRepository.java:
```diff
@@ -0,0 +1,37 @@
+	 * @param id
+	 * @return Check {@link com.revature.services.impl.TripServiceImpl}
+	 */
+	@Query("select t from Trip t where t.driver.userId = ?1")
```
Have either of these queries been tested? I am also curious why the second query returns a list of the trip ids, instead of just a list of the Trip objects.

flacks replied:
Updated to return Trip objects. Makes things a lot simpler now! Thank you very much for your input there, too!

IkenoXamos commented on src/main/java/com/revature/services/impl/TripServiceImpl.java:
```diff
@@ -0,0 +1,57 @@
+
+	@Override
+	public Trip getTripById(int id) {
+		return tripRepository.findById(id).get();
```
I see that most other services structure the getById method the same way, but we should consider refactoring them. If this ever gets called with an id that does not exist, it will throw an exception, which is not ideal.

flacks replied:
I agree we should refactor every method that could throw an exception. For now, refactored only TripRepository's findById to return an Optional.

IkenoXamos commented on src/main/java/com/revature/services/impl/TripServiceImpl.java:
```diff
@@ -0,0 +1,57 @@
+		List<Integer> tripIds = tripRepository.getTripsByRiderId(riderId);
+		List<Trip> trips = null;
+		for (int r : tripIds) {
+			trips.add(tripRepository.findById(r).get());
```
Same issue can occur here, if any of the ids do not exist in the database, an exception will be thrown.

flacks replied:
Refactored.

Jcfo233 committed:
added date to trip model

flacks committed:
Trip updates

- Properly handle not overwriting Trips on POST if they exist
- Properly handle not saving Trips if they don't exist on PUT
- Fix typo in Javadoc for deleting trips by ID
- Use @Data annotation instead of many other Lombok annotations
- Update Trip model ID column name, add departure address, rename
  destination column, add tripDate
- Remove trips field from the User model
- Properly implement and test getTripsByRiderId in TripRepository

Co-authored-by: The Camel <jcfo233@gmail.com>
Co-authored-by: Jose Avalos <jlavalos73@gmail.com>
Co-authored-by: Galo Romero <grome014@fiu.edu>
Co-authored-by: Gilbert Turner <grturner@5x5code.com>

flacks committed:
Remove unneeded dependencies

flacks committed:
Merge branch 'dev' into trips

flacks committed:
Remove redundant access modifiers

flacks committed:
Tidy indentation in pom.xml

flacks committed:
Actually use the return value of updateTrip

IkenoXamos commented on src/main/java/com/revature/repositories/TripRepository.java:
```diff
@@ -0,0 +1,37 @@
+	 * @param id
+	 * @return Check {@link com.revature.services.impl.TripServiceImpl}
+	 */
+	@Query(value = "select t.TRIP_ID, t.AVAILABLE_SEATS, t.NAME, t.DEPARTURE_ID, t.DESTINATION_ID, t.USER_ID, t.TRIP_DATE from trips t join riders r on t.TRIP_ID = r.TRIP_ID where RIDER_ID = :id", nativeQuery = true)
```
If you want to return the entire trip, you can simply use "select t from ..."

flacks replied:
Resolved. Used `*` instead of `t` since this is a native query.

IkenoXamos commented on src/main/java/com/revature/models/Trip.java:
```diff
@@ -0,0 +1,51 @@
+    @NotNull
+    @ManyToOne
+    @JoinColumn(name = "departure_id")
+    private Address departure;
```
These two address fields should also have @ Valid.

flacks replied:
Added.

flacks committed:
Touch-ups

- Add @Valid annotations to address fields in Trip model
- Simplify native SQL query for getTripsByRiderId

Co-authored-by: The Camel <jcfo233@gmail.com>
Co-authored-by: Jose Avalos <jlavalos73@gmail.com>
